design/providerid_format_change.md

```diff
@@ -0,0 +1,173 @@
+
+Also, since it is constructed in the code, there is no need for the users to
+provide provider-id and label via templates. However, the users still refer to
+the providerID variable in their templates to get it populated by cloud-init.
+
+### User Stories
+
+- As a user, I want my nodes' providerIDs to be unique per kubernetes deployment
+ and be generated automatically by CAPM3 using metal3machine and baremetalhost resource names.
+
+## Design Details
+
+The following changes are required in the code:
+
+- Searching nodes is no longer done using labels
+- Searching nodes is done using implicitly set providerid
+- Providerid field is introduced in metal3datatemplate in the code.
+- Labels are no longer provided via templates
+
+Concerning backwards compatibility, there are three scenarios:
+
+1. backwards compatibility with legacy providerID in existing deployments,
+  - During scale out, new nodes keep using legacy providerID
+  - During upgrade, upgraded nodes keep using legacy providerID
+2. backwards compatibility with legacy providerID for new deployments,
+  - During provisioning, if user supplies template with legacy providerID and
+  node label, then the nodes should be provisioned with the legacy providerID
+  - This will be supported for the next two releases
+
+3. transition from legacy to new providerID format for existing deployments.
+  - Existing nodes with legacy providerID can be moved to the new format using
+  upgrade. This can be achieved by creating new
+  `kubeadmconfigtemplates.bootstrap.cluster.x-k8s.io` resource with the new provider-id
+  set inside the `.spec.template.spec.joinConfiguration.nodeRegistration.kubeletExtraArgs`
+  structure. An example is shown below:
+
+```yaml
+apiVersion: bootstrap.cluster.x-k8s.io/v1beta1
+kind: KubeadmConfigTemplate
+metadata:
+  name: test1-s54fx
+  namespace: metal3
+spec:
+  clusterConfiguration: ....
+  files: ...
+  joinConfiguration:
+    controlPlane:
+      localAPIEndpoint: {}
+    discovery: ...
+    nodeRegistration:
+      kubeletExtraArgs:
+        provider-id: metal3://{{ ds.meta_data.providerid }}
+      name: '{{ ds.meta_data.name }}'
+```
+
+### Implementation Details/Notes/Constraints
+
+None
+
+### Risks and Mitigations
+
+None
+
+### Work Items
+
+None
+
+### Dependencies
+
+- Templates should be updated with the new format
+
+### Test Plan
+
+Existing upgrade tests can serve for verification of the new format. This includes
+upgrading with templates that use the new providerID format.
+
+### Upgrade / Downgrade Strategy
+
+The code change on the providerID logic will be backward compatible
+in that CAPM3 keeps populating the providerID from the label, if the providerID
+is unset and the label is set.
+
+### Version Skew Strategy
+
+None
+
+## Drawbacks
+
+None
+
+## Alternatives
+
+There are possible formats:
+
+1. ```metal3://<namespace>/<BareMetalHost name>/<metal3machine name>```
```

Comment:
so.. since metal3machine is not actually a property of the provider, maybe

```metal3://<namespace>/<BareMetalHost name>?owner=<metal3machine name>```

What I'm trying to avoid is the users guessing which part of the URL is BMH name.